pkg/arvo/tests/sys/zuse/crypto/ed25519.hoon

```diff
@@ -0,0 +1,156 @@
+::  tests for the ed25519 signature algorithm
+::
+/+  *test
+=,  ed:crypto
+::
+|%
+::
+::  Test vectors from Section 7.1 of RFC 8032:
+::  https://tools.ietf.org/html/rfc8032#section-7.1
+::
++$  vector  [seed=@ux pk=@ux msg=@ux sig=@ux]
+::
+++  test-vectors
+  ^-  tang
+  |^  ;:  weld
+        %+  category  "puck"
+        (zing (turn vectors check-puck))
+        %+  category  "sign"
+        (zing (turn vectors check-sign))
+        %+  category  "veri"
+        (zing (turn vectors check-veri))
+      ==
+  ::
+  ++  check-puck
+    |=  vector
+    %+  expect-eq
+      !>  pk
+      !>  `@ux`(puck seed)
+  ::
+  ++  check-sign
+    |=  vector
+    %+  expect-eq
+      !>  sig
+      !>  `@ux`(sign msg seed)
+  ::
+  ++  check-veri
+    |=  vector
+    %+  expect-eq
+      !>  %.y
+      !>  (veri sig msg pk)
+  ::
+  ++  vectors
+    (turn vectors-raw swp-vec)
+  ++  swp-vec
+    |=  vector
+    :^  (swp 3 seed)
+        (swp 3 pk)
+        (swp 3 msg)
+        (swp 3 sig)
```

Comment:
this feels ugly; how can I clean this up? I want to map `(cury swp 3)` over each element of each vector.

btw, it took me a while to realize that my tests were failing because I wasn't correcting for endianness...I guess it's more convenient to operate on curves this way, but it definitely tripped me up. *Especially* because `shan` and `shaz` have different endiannness...es. I thought I was going crazy when the `shan` output was clearly correct and the `shaz` output was not. 😅

Reply:
@lukechampine, it is kind of ugly, but there's no general, typed way to map over arbitrary-width tuples (You could do it over raw nouns, although i'd advice against it.)

Little-endian is the right default for hoon atom operations, including crypto IMO, but the inconsistencies you mentioned are really bad. It'd be nice to get them cleaned up.

Reply:
Personally I'd just collapse to one line and call it a day, it's not _that_ much code, but:

ye of little faith, check out ++scanf and ++raid in zuse. If what you have is a fixed-length tuple of the shape `|-($@(@ [@ $]))`, then a wet gate something like `|*(a=|-($@(@ [@ $])) ^+(a ?@(a (swp 3 a) [(swp 3 -.a) (..$ +.a)])))` will monomorphize itself repeatedly for the different tuple lengths, producing a tuple of the same width at each step. (in this case same exact type even, hence `^+ a`.) 

To reiterate, for just four elements the explicit version is much more readable

```
[(swp 3 seed) (swp 3 pk) (swp 3 msg) (swp 3 seed)]
```

or if you must

```
=/  f  |=(a=@ (swp 3 a))
[(f seed) (f pk) (f msg) (f seed)]
```

Reply:
Very nice! Looks like `(..$ +.a)` is what makes this go, forcing the re-monomorphization.

It requires a little casting to make it work (and a `$;` for the manual mold construction):

```
> =swp-tup |*(a=$;(|-($@(@ [@ $]))) ^+(a ?@(a (swp 3 a) [(swp 3 -.a) (..$ +.a)])))
> (swp-tup `@`%foo)
6.713.199
> (swp-tup `@`%foo `@`%bar)
[6.713.199 6.447.474]
> (swp-tup `@`%foo `@`%bar `@`%baz)
[6.713.199 6.447.474 6.447.482]
```

Reply:
`(swp 3 %foo)` does not have the value `%foo` and so cannot be cast back to its constant type, indeed :P

Reply:
I'd recommend e.g. `@ta` to verify the output clearer; and prefix `? ` since the type being right is the important part

Reply:
```
> =swp-tup |*(a=$;(|-($@(@ [@ $]))) ^+(a ?@(a (swp 3 a) [(swp 3 -.a) (..$ +.a)])))
> ? (swp-tup ~.foobarbaz)
  @ta
~.zabraboof
> ? (swp-tup ~.foo ~.bar ~.baz)
  {@ta @ta @ta}
[~.oof ~.rab ~.zab]
```